File: asp/java/Heap.java
public class Heap {
    int Cap;   // maximum number of elements
    boolean Min;  // max-heap if false, min-heap if true
    int Len;   // current number of elements
    int[] Buf; // backing array, length = capacity

    public Heap(int capy, boolean mini) {
        this.Cap = capy;
        this.Min = mini;
        this.Len = 0;
        this.Buf = new int[capy];
        for (int i = 0; i < capy; i++) {
            this.Buf[i] = 0;
        }
    }

    public void Reset() {
        this.Len = 0;
        for (int i = 0; i < this.Cap; i++) {
            this.Buf[i] = 0;
        }
    }   

    public int Top() {
        if (this.Len == 0) {
            throw new IllegalStateException("Heap.Top on empty heap");
        }
        return this.Buf[0];
    }

    public void Push(int x) {
        if (this.Len == this.Buf.length) {
            throw new IllegalStateException("Heap overflow");
        }
        this.Buf[this.Len] = x;
        this.up(this.Len);
        this.Len++;
    }

    public int Pop() {
        if (this.Len == 0) {
            throw new IllegalStateException("Heap.Pop on empty heap");
        }
        int top = this.Buf[0];
        this.Len--;
        this.swap(0, this.Len);
        this.down(0);
        return top;
    }

    private boolean cmp(int i, int j) {
        int ai = this.Buf[i];
        int aj = this.Buf[j];
        if (this.Min) {
            return ai < aj;
        } else {
            return ai > aj;
	    }
    }

    private void swap(int i, int j) {
        int temp = this.Buf[i];
        this.Buf[i] = this.Buf[j];
        this.Buf[j] = temp;
    }

    private void up(int i) {
        while (true) {
            if (i == 0) {
                return;
            }
            int p = (i - 1) >> 1;
            if (!this.cmp(i, p)) {
                return;
            }
            this.swap(i, p);
            i = p;
        }
    }

    private void down(int i) {
        while (true) {
            int l = i*2 + 1;
            if (l >= this.Len) {
                return;
            }
            int j = l;
            int r = l + 1;
            if (r < this.Len && this.cmp(r, l)) {
                j = r;
            }
            if (!this.cmp(j, i)) {
                return;
            }
            this.swap(i, j);
            i = j;
        }
    }

    // Remove the deepest occurrence of val from the heap.
    protected boolean remove(int val) {
        for (int i = this.Len - 1; i >= 0; i--) {
            if (this.Buf[i] == val) {
                this.removeAt(i);
                return true;
            }
        }
        return false;
    }

    // Remove the element at index i from the heap.
    private int removeAt(int i) {
        if (i == 0) {
            return this.Pop();
        }
        if (i >= this.Len) {
            throw new IllegalArgumentException("invalid index");
        }
        int r = this.Buf[i];
        this.Len--;
        this.swap(i, this.Len);
        // Try both up and down since the swapped element
        // may violate heap property in either direction.
        this.up(i);
        this.down(i);
        return r;
    }
}   

File: asp/gol/heap.go
package gol

// Fixed-capacity heap.
type Heap struct {
	Cap int   // maximum number of elements
	Min bool  // max-heap if false, min-heap if true
	Len int   // current number of elements
	Buf []int // backing array, length = capacity
}

func NewHeap(capy int, mini bool) *Heap {
	if capy <= 0 {
		panic("Heap capacity must be > 0")
	}
	return &Heap{
		Cap: capy,
		Min: mini,
		Len: 0,
		Buf: make([]int, capy),
	}
}

func (h *Heap) Reset() {
	h.Len = 0
	for i := range h.Buf {
		h.Buf[i] = 0
	}
}

func (h *Heap) Top() int {
	if h.Len == 0 {
		panic("Heap.Top on empty heap")
	}
	return h.Buf[0]
}

func (h *Heap) Push(x int) {
	if h.Len == len(h.Buf) {
		panic("Heap overflow")
	}
	h.Buf[h.Len] = x
	h.up(h.Len)
	h.Len++
}

func (h *Heap) Pop() int {
	if h.Len == 0 {
		panic("Heap.Pop on empty heap")
	}
	top := h.Buf[0]
	h.Len--
	h.swap(0, h.Len)
	h.down(0)
	return top
}

func (h *Heap) cmp(i, j int) bool {
	ai, aj := h.Buf[i], h.Buf[j]
	if h.Min {
		return ai < aj
	} else {
		return ai > aj
	}
}

func (h *Heap) swap(i, j int) {
	h.Buf[i], h.Buf[j] = h.Buf[j], h.Buf[i]
}

func (h *Heap) up(i int) {
	for {
		if i == 0 {
			return
		}
		p := (i - 1) >> 1
		if !h.cmp(i, p) {
			return
		}
		h.swap(i, p)
		i = p
	}
}

func (h *Heap) down(i int) {
	for {
		l := i*2 + 1
		if l >= h.Len {
			return
		}
		j := l
		r := l + 1
		if r < h.Len && h.cmp(r, l) {
			j = r
		}
		if !h.cmp(j, i) {
			return
		}
		h.swap(i, j)
		i = j
	}
}

// Remove the deepest occurrence of val from the heap.
func (h *Heap) remove(val int) bool {
	for i := h.Len - 1; i >= 0; i-- {
		if h.Buf[i] == val {
			_ = h.removeAt(i)
			return true
		}
	}
	return false
}

// Remove the element at index i from the heap.
func (h *Heap) removeAt(i int) int {
	if i == 0 {
		return h.Pop()
	}
	if i >= h.Len {
		panic("invalid index")
	}
	r := h.Buf[i]
	h.Len--
	h.swap(i, h.Len)
	// Try both up and down since the swapped element
	// may violate heap property in either direction.
	h.up(i)
	h.down(i)
	return r
}
